src/definitions.rs

```diff
@@ -0,0 +1,306 @@
+                    || (!f.metadata.name.contains('$')
+                        && !is_valid_function_name(&f.metadata.name)),
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+impl FuncInfo {
+    fn write_declaration(&self, writer: &mut dyn fmt::Write, operator: bool) -> fmt::Result {
+        for comment in &*self.metadata.comments {
+            writeln!(writer, "{comment}")?;
+        }
+
+        if operator {
+            writer.write_str("op ")?;
+        } else {
+            writer.write_str("fn ")?;
+        }
+
+        if let Some(name) = self.metadata.name.strip_prefix("get$") {
+            write!(writer, "get {name}(")?;
+        } else if let Some(name) = self.metadata.name.strip_prefix("set$") {
+            write!(writer, "set {name}(")?;
+        } else {
+            write!(writer, "{}(", self.metadata.name)?;
+        }
+
+        let mut first = true;
+        for i in 0..self.metadata.params {
+            if !first {
+                writer.write_str(", ")?;
+            }
+            first = false;
+
+            let (param_name, param_type) = self
+                .metadata
+                .params_info
+                .get(i)
+                .map(|s| {
+                    let mut s = s.split(':');
+                    (
+                        s.next().unwrap_or("_").split(' ').last().unwrap(),
+                        s.next().map(decl_type_name).unwrap_or(Cow::Borrowed("?")),
+                    )
+                })
+                .unwrap_or(("_", "?".into()));
+
+            if operator {
+                write!(writer, "{param_type}")?;
+            } else {
+                write!(writer, "{param_name}: {param_type}")?;
+            }
+        }
+
+        write!(
+            writer,
+            ") -> {};",
+            decl_type_name(&self.metadata.return_type)
+        )?;
+
+        Ok(())
+    }
+}
+
+/// We have to transform some of the types.
+///
+/// This is highly inefficient and is currently based on
+/// trial and error with the core packages.
+///
+/// It tries to flatten types, removing `&` and `&mut`,
+/// and paths, while keeping generics.
+///
+/// Associated generic types are also rewritten into regular
+/// generic type parameters.
+fn decl_type_name(ty: &str) -> Cow<str> {
+    let ty = ty.replace("crate::", "");
+    let ty = ty.split("::").last().unwrap();
+    let ty = ty.trim();
+    let ty = ty.strip_prefix("&mut").unwrap_or(ty).trim();
+
+    let ty = ty
+        .strip_prefix("RhaiResultOf<")
+        .and_then(|s| s.strip_suffix('>'))
+        .map(str::trim)
+        .unwrap_or(ty);
+
+    // FIXME(tamasfe): some of the given types are unusable,
+    // e.g. "Range<crate" is the entire type this function receives,
+    // same with "crate", I have no idea where these come from.
+    ty.replace("Range<crate", "?")
+        .replace("crate", "?")
+        .replace("Iterator<Item=", "Iterator<")
+        .replace("Dynamic", "?")
+        .replace("INT", "int")
```

Comment:
Thanks, unfortunately it's not enough, I still need to do some plumbing afterwards.

Reply:
Beware, though, that some people may register functions with `i64`, say, instead of `INT`, so the signature will not have `int` there.  Perhaps you want to also map `std::any::type_name::<crate::INT>()` and `std::any::type_name::<crate::FLOAT>()` back to `int` and `float` to be consistent.

Reply:
Last time I checked even `i64` and `i32` were incompatible in Rhai (as in no implicit cast), so I guess it's fine to make these types distinct.